rfcs/20240312-standardize-tan-customcall-tuplecollective-ops.md

```diff
@@ -0,0 +1,69 @@
+# [RFC] Standardize TanOp, CustomCallOp with typed FFI, and tuple-collectives ops
```

Comment:
Opening comment here since GH replies are difficult to keep track of:

@loganchien:
It seems that there are 3 different features proposed in this RFC. And they look unrelated to each other IMHO. May you split this into 3 RFCs?

@mlevesquedion:
I agree with comments from others that maybe this should be 3 separate RFCs? Seems like these are independent changes.

Reply:
The common theme here is "Standardize successful MHLO features into StableHLO" -- aka HLO/MHLO/StableHLO parity for these features. In that regard we could split that into 3 separate RFCs, but there would be some duplication of motivation. I'm on board with either.

Reply:
I suggested to split this into 3 RFCs because they have different complexity.  Some portion can be merged faster.  Some need thorough review.

`TanOp` is trivial.  IMO, it can be on fast track.  I don't see the reason not merging it ASAP.  However, in the RFC, we should mention how to maintain forward/backward compatibility for `TanOp` (p.s. it  should be trivial as lowering `TanOp(x)` to `SinOp(x) / CosOp(x)` for backward compatibility).

I am supportive to the `CustomCallOp` change, but I would like to see more elaboration on rationale and refinement on wordings.  For example, I think instead of copying HLO terminology or define spec as existing implementation, it will be better to focus the wording on the properties of the construct.  Instead of using "FFI", we can just describe the feature as "extend `backend_config` to take a `DictionaryAttr`", etc.

For the collective ops, I don't have any idea.